### PR TITLE
Compatibility with postgresql 12

### PR DESCRIPTION
**Pull request by CyberDem0n**
* use `SHOW primary_conninfo` instead of parsing config file on pg12
* parse the value of primary_conninfo in order to compare parameters instead of doing a comparison of raw string
* cache parse results until the next change of config file (track mtime)
* strip out standby and recovery parameters from postgresql.auto.conf before starting the postgres 12
* simplify ConfigHandler.reload_config: the new compare_values function handles all corner cases for us

Patroni config remains backward compatible.
Despite for example `restore_command` converted to a GUC starting from postgresql 12, in the Patroni configuration you can still keep it in the `postgresql.recovery_conf` section.
If you put it into `postgresql.parameters.restore_command`, that will also work, but it is important not to mix both ways:
```yaml
# is OK
postgresql:
  parameters:
    restore_command: my_restore_command
    archive_cleanup_command: my_archive_cleanup_command

# is OK
postgresql:
  recovery_conf:
    restore_command: my_restore_command
    archive_cleanup_command: my_archive_cleanup_command

# is NOT ok
postgresql:
  parameters:
    restore_command: my_restore_command
  recovery_conf:
    archive_cleanup_command: my_archive_cleanup_command
```

### CODE REVIEW

File: patroni/postgresql/config.py [hunk 1]
```diff
@@ -6,19 +6,149 @@
...
 SYNC_STANDBY_NAME_RE = re.compile(r'^[A-Za-z_][A-Za-z_0-9\$]*$')
+PARAMETER_RE = re.compile(r'([a-z_]+)\s*=\s*')
 
 
 def quote_ident(value):
     """Very simplified version of quote_ident"""
     return value if SYNC_STANDBY_NAME_RE.match(value) else '"' + value + '"'
 
 
+def conninfo_uri_parse(dsn):
```
[the remark this answers is not in the document]
Comment (CyberDem0n) [reply patroni/postgresql/config.py]: They logically belong only to this part and not used anywhere else. There is not much value to create a separate file for them.

File: patroni/postgresql/config.py [hunk 2]
```diff
@@ -6,19 +6,149 @@
...
+        ret = conninfo_parse(value)
+
+    if ret:
+        requiressl = ret.pop('requiressl', None)
+        if requiressl == '1':
+            ret['sslmode'] = 'require'
+        elif requiressl is not None:
+            ret['sslmode'] = 'prefer'
+        if 'dbname' in ret:
+            del ret['dbname']
```
[the remark this answers is not in the document]
Comment (CyberDem0n) [reply patroni/postgresql/config.py]: The database name is never used for physical replication (in the primary_conninfo).

File: patroni/postgresql/config.py [hunk 3]
```diff
@@ -6,19 +6,149 @@
...
+    return ret
+
+
+def mtime(filename):
+    try:
+        st = os.stat(filename)
+        return st.st_mtime
+    except OSError:
+        return 0
```
[the remark this answers is not in the document]
Comment (CyberDem0n) [reply patroni/postgresql/config.py]: Nothing. We just want to check that mtime didn't change. Maybe None would be slightly better.

**Comment by erthalion** [comment]
:+1:

**Comment by CyberDem0n** [comment]
:+1:

**Pull request merged**